scvi/external/scar/_model.py

```diff
@@ -0,0 +1,334 @@
+        %(param_size_factor_key)s
+        """
+        setup_method_args = cls._get_setup_method_args(**locals())
+        anndata_fields = [
+            LayerField(REGISTRY_KEYS.X_KEY, layer, is_count_data=True),
+            CategoricalObsField(REGISTRY_KEYS.BATCH_KEY, batch_key),
+            CategoricalObsField(REGISTRY_KEYS.LABELS_KEY, labels_key),
+            NumericalObsField(
+                REGISTRY_KEYS.SIZE_FACTOR_KEY, size_factor_key, required=False
+            ),
+        ]
+        adata_manager = AnnDataManager(
+            fields=anndata_fields, setup_method_args=setup_method_args
+        )
+        adata_manager.register_fields(adata, **kwargs)
+        cls.register_manager(adata_manager)
+
+    @staticmethod
+    def get_ambient_profile(
+        adata: AnnData,
+        raw_adata: AnnData,
+        prob: float = 0.995,
+        min_raw_counts: int = 2,
+        iterations: int = 3,
+        n_batch: int = 1,
+        sample: int = 50000,
+    ):
+        """Calculate ambient profile for relevant features
+        Identify the cell-free droplets through a multinomial distribution. See EmptyDrops [Lun2019]_ for details.
+
+        Parameters
+        ----------
+        adata : AnnData
+            A filtered adata object, loaded from filtered_feature_bc_matrix using `scanpy.read`, gene filtering is recommended to save memory
+        raw_adata : AnnData
+            A raw adata object, loaded from raw_feature_bc_matrix using `scanpy.read`
+        prob : float, optional
+            The probability of each gene, considered as containing ambient RNA if greater than prob (joint prob euqals to the product of all genes for a droplet), by default 0.995
+        min_raw_counts : int, optional
+            Total counts filter for raw_adata, filtering out low counts to save memory, by default 2
+        iterations : int, optional
+            Total iterations, by default 3
+        n_batch : int, optional
+            Total number of batches, set it to a bigger number when out of memory issue occurs, by default 1
+        sample : int, optional
+            Randomly sample droplets to test, if greater than total droplets, use all droplets, by default 50000
+
+        Returns
+        -------
+        The relevant ambient profile is added in `adata.varm`
+        """
+        # take subset genes to save memory
+        raw_adata = raw_adata[:, raw_adata.var_names.isin(adata.var_names)]
+        raw_adata = raw_adata[raw_adata.X.sum(axis=1) >= min_raw_counts]
+
+        sample = int(sample)
+        idx = np.random.choice(
+            raw_adata.shape[0], size=min(raw_adata.shape[0], sample), replace=False
+        )
+        raw_adata = raw_adata[idx]
+        print(f"Randomly sampling {sample} droplets to calculate the ambient profile.")
+        # initial estimation of ambient profile, will be updated
+        ambient_prof = raw_adata.X.sum(axis=0) / raw_adata.X.sum()
+
+        for _ in tqdm(range(iterations)):
```

Comment:
scvi.utils.track is a wrapper on tqdm and rich I'm trying to use codebase wide

Reply:
done in latest commit